snap/quota/resources_builder.go

```diff
@@ -0,0 +1,108 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+* Copyright (C) 2022 Canonical Ltd
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License version 3 as
+* published by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+ */
+
+package quota
+
+import (
+	"github.com/snapcore/snapd/gadget/quantity"
+)
+
+type ResourcesBuilder struct {
+	MemoryLimit    quantity.Size
+	MemoryLimitSet bool
+
+	CPUCount    int
+	CPUCountSet bool
+
+	CPUPercentage    int
+	CPUPercentageSet bool
+
+	AllowedCPUs    []int
+	AllowedCPUsSet bool
+
+	ThreadLimit    int
+	ThreadLimitSet bool
+}
+
+func (rb *ResourcesBuilder) WithMemoryLimit(limit quantity.Size) *ResourcesBuilder {
+	rb.MemoryLimit = limit
+	rb.MemoryLimitSet = true
+	return rb
+}
+
+func (rb *ResourcesBuilder) WithCPUCount(count int) *ResourcesBuilder {
+	rb.CPUCount = count
+	rb.CPUCountSet = true
+	return rb
+}
+
+func (rb *ResourcesBuilder) WithCPUPercentage(percentage int) *ResourcesBuilder {
+	rb.CPUPercentage = percentage
+	rb.CPUPercentageSet = true
+	return rb
+}
+
+func (rb *ResourcesBuilder) WithAllowedCPUs(allowedCPUs []int) *ResourcesBuilder {
+	rb.AllowedCPUs = allowedCPUs
+	rb.AllowedCPUsSet = true
+	return rb
+}
+
+func (rb *ResourcesBuilder) WithThreadLimit(limit int) *ResourcesBuilder {
+	rb.ThreadLimit = limit
+	rb.ThreadLimitSet = true
+	return rb
+}
+
+func (rb *ResourcesBuilder) Build() Resources {
```

Comment:
Perhaps I should have commented earlier in the main PR, but I think this went a bit too much in the usual builder pattern. If you look at the code below, it's not really doing anything useful aside from moving fields from one struct to the other, so perhaps it's enough to have:
```
func (r *Resources) WithThreadLimit(..) *Resources { .. }
func (r *Resources) WithAllowedCPUs(..) *Resources { .. }
func (r *Resources) WithCPUCount(..) *Resources { .. }
func (r *Resources) WithMemoryLimit(..) *Resources { .. }
```
which all just annotate the existing `r` and return it, so we can chain the calls like `NewResources().WithThis().WithThat()`, but we don't really need the builder to copy the things around and overall looks more like what you expect of Go code

Reply:
Ok so @bboozzoo, I'll just shortly reiterate our conversion, my personal opinion on this is that it's nice to have the builder as it's primarily used by tests cases all around to build resource structures, and it's a lot cleaner to have that construction logic seperated from the resources object as it's only a single place (quota.go:123) the builder is actually used in the code, and that is just for building the resources object when converting between structure formats. 

@bboozzoo then pointed out maybe the builder should live somewhere else and we should just construct the resources object explicitly where it's used, which I think is fine, but for lack of better place to put it we've left it where it is, unless someone else has some input of whether the builder is fine to have where it is or it should be moved to it's own test package.

Reply:
With* methods do usually make a copy in go, that would be wasteful in a different way. I find the builder slightly overkill outside of tests but don't have a strong opinion to move it.